### PR TITLE
Implement Sync Versioned Transition logic

### PR DESCRIPTION
**Pull request by xwduan**
## What changed?

Implement Sync Versioned Transition logic
## How did you test it?

test will be added later.

### CODE REVIEW

File: service/history/ndc/workflow_state_replicator.go [hunk 1]
```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
...
+			historyBlob, err := remoteHistoryIterator.Next()
+			if err != nil {
+				return err
+			}
+
+			txnID, err := r.shardContext.GenerateTaskID()
+			if err != nil {
+				return err
+			}
+			_, err = r.executionMgr.AppendRawHistoryNodes(ctx, &persistence.AppendRawHistoryNodesRequest{
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/ndc/workflow_state_replicator.go:661]: We can also go with existing approach: gather missing events in memory, then an immutable history builder to mutable state with those events. Then executionStats update will be handled by persistence execution manager.

File: service/history/ndc/workflow_state_replicator.go [hunk 2]
```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
...
+			namespaceId,
+			workflowId,
+			runId,
+			startId,
+			startVersion,
+			endId,
+			endVersion),
+		)
+		for remoteHistoryIterator.HasNext() {
+			historyBlob, err := remoteHistoryIterator.Next()
```
Comment (yycptt) [inline service/history/ndc/workflow_state_replicator.go:652]: Is one blob one history batch? How many persistence requests will be made in existing resender impl v.s this impl.

File: service/history/ndc/workflow_state_replicator.go [hunk 3]
```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
...
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return newRunMutableState, nil
+}
+
+// this function does not handle reset case, so it should only be used for the case where workflow run is found at local cluster
+func (r *WorkflowStateReplicatorImpl) bringLocalEventsUpToSourceCurrentBranch(
```
Comment (yycptt) [inline service/history/ndc/workflow_state_replicator.go:555]: for future: we may need to some checkpoint mechanism for backfilling to handle large histories. One idea can be: create a temp branch in version histories and use that to record how many events have been backfilled.

File: service/history/ndc/workflow_state_replicator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/ndc/workflow_state_replicator.go:752]: Added logic to add size.

File: service/history/ndc/workflow_state_replicator.go (hunk 2 above)
Comment (xwduan) [reply service/history/ndc/workflow_state_replicator.go:743]: Yeah, seems this is also used by current resend. The current closed workflow replication back fill logic is also using the same logic.

 I did not find an API that writes multiple batches of history events.

File: service/history/ndc/workflow_state_replicator.go [hunk 4]
```diff
@@ -187,7 +198,551 @@ func (r *WorkflowStateReplicatorImpl) SyncWorkflowState(
...
+			workflowId,
+			runId,
+			nil,
+		)
+	}
+	if localMutableState == nil {
+		return r.applySnapshotWhenWorkflowNotExist(ctx, namespaceId, workflowId, runId, wfCtx, releaseFn, request.Snapshot, request.SourceClusterName, request.NewRunInfo)
+	}
+	var isBranchSwitched bool
+	localTransitionHistory := workflow.CopyVersionedTransitions(localMutableState.GetExecutionInfo().TransitionHistory)
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/ndc/workflow_state_replicator.go]: Updated the code to compare history events to decide if the branch is changed when local transition history is empty.

File: service/history/shard/engine.go [hunk 5]
```diff
@@ -163,4 +163,13 @@ type (
 		NewEvents []*historypb.HistoryEvent
 		NewRunID  string
 	}
+
+	ReplicateMutableStateRequest struct {
+		Mutation                          *persistencespb.WorkflowMutableStateMutation
+		Snapshot                          *persistencespb.WorkflowMutableState
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/shard/engine.go:169]: This type of structure is been reuse/created multiple times. I will consolidate them in a separate pr.

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 6]
```diff
@@ -237,4 +237,17 @@ message VerifyVersionedTransitionTaskAttributes {
 message SyncVersionedTransitionTaskAttributes {
     repeated temporal.api.common.v1.DataBlob event_batches = 1;
     NewRunInfo new_run_info = 2;
+    oneof state_attributes {
+        SyncWorkflowStateMutationAttributes sync_workflow_state_mutation_attributes = 3;
+        SyncWorkflowStateSnapshotAttributes sync_workflow_state_snapshot_attributes = 4;
+    }
+}
+
+message VersionedTransitionArtifact {
```
Comment (xwduan) [inline proto/internal/temporal/server/api/replication/v1/message.proto:246]: I will use this struct to replace the above one in next pr.

**Review by hai719 — APPROVED** [review]

**Pull request merged**